upf/upf_pfcp_api.c

```diff
@@ -799,9 +805,163 @@ lookup_nwi (u8 * name)
   return pool_elt_at_index (gtm->nwis, p[0]);
 }
 
+static bool
+validate_teid (u32 teid)
+{
+  return ((teid != 0) && (teid != 0xffffffff));
+}
+
+static int
+teid_v4_lookup (upf_main_t * gtm, u32 teid, upf_upip_res_t * res)
+{
+  clib_bihash_kv_8_8_t kv, value;
+  gtpu4_tunnel_key_t key4;
+
+  key4.dst = res->ip4.as_u32;
+  key4.teid = teid;
+
+  kv.key = key4.as_u64;
+
+  if (PREDICT_FALSE
+      (clib_bihash_search_8_8 (&gtm->v4_tunnel_by_key, &kv, &value)))
+    return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+static int
+teid_v6_lookup (upf_main_t * gtm, u32 teid, upf_upip_res_t * res)
+{
+  clib_bihash_kv_24_8_t kv, value;
+
+  kv.key[0] = res->ip6.as_u64[0];
+  kv.key[1] = res->ip6.as_u64[1];
+  kv.key[2] = teid;
+
+  if (PREDICT_FALSE
+      (clib_bihash_search_24_8 (&gtm->v6_tunnel_by_key, &kv, &value)))
+    return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+static u32
+process_teid_generation (upf_main_t * gtm, u8 chid, u32 flags,
+			 upf_upip_res_t * res, upf_session_t * sx)
+{
+  u8 retry_cnt = 10;
+  u32 teid = 0;
+  bool ok = false;
+  do
+    {
+      teid = random_u32 (&gtm->rand_base) & (UINT32_MAX ^ res->mask);
+      if (!validate_teid (teid))
+	{
+	  retry_cnt--;
+	  continue;
+	}
+
+      gtm->rand_base = teid;
+
+      ok =
+	((flags & F_TEID_V4) || (flags & F_TEID_V6)) &&
+	((teid_v4_lookup (gtm, teid, res) != 0) &&
+	 ((teid_v6_lookup (gtm, teid, res) != 0)));
```

Comment:
I gave an example for this in https://github.com/travelping/upg-vpp/pull/3#discussion_r506163343.

What made you think that your version is correct or semantically equivalent to my suggestion?

```suggestion
      ok =
	(!(flags & F_TEID_V4) || teid_v4_lookup (gtm, teid, res) == 0) &&
	(!(flags & F_TEID_V6) || teid_v6_lookup (gtm, teid, res) == 0);
```

Reply:
Way I've done is kind of a an obvious expression of desired logic
OK = false if teid lookup is failed for any of given teid+endpoint ip combination regardless it's version. And true otherwise.

If you will take a look at teid_v4/6_lookup function, it returns UPF_GTPU_ERROR_NO_SUCH_TUNNEL if there is no tunnel with given teid/res, 0 if there is one

I was having doubts about logic you've introduced. Let's will pick up simplest scenario with V4 flag and there is no tunnel created for given teid, and for simplified pipeline let's call (flags & F_TEID_V4) as 1(basically every int not equal to 0), and UPF_GTPU_ERROR_NO_SUCH_TUNNEL as -1 without breaking general logic
so we have
(!1 || -1 == 0) && (!0 || -1 ==0)
which as a result gives us 0 for any of teid generated. However, we do expecting 1 here cause V4 flag is given and no tunnel is yet created for given combination of teid and endpoint ip

Reply:
Logic you've introduced is fine if `==` change to `!=`, probably you've just missed return value of teid lookup.
I've changed this and left a comment, that teid lookup for both ip4/ip6 is safe even there is no IP4/IP6 address configured for upip_res.

Reply:
I don't know why you need that comment. The lookup will only be invoked if corresponding flag is set.

But it doesn't hurt to have it, so no need to remove it.